src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoEnumTrailingCommaCheck.java

```diff
@@ -0,0 +1,129 @@
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ *  Checks that enum definition does not contain a trailing comma.
+ *  Rationale: JLS allows trailing commas in arrays and enumerations, but does not allow
+ *  them in other locations. To unify the coding style, the use of trailing commas should
+ *  be prohibited.
+ * </p>
+ * <pre>
+ * enum Foo1 {
+ *   FOO,
+ *   BAR;
+ * }
+ * </pre>
+ * <p>
+ *  The check demands that there should not be any comma after last constant in
+ *  enum definition.
+ * </p>
+ * <pre>
+ * enum Foo1 {
+ *   FOO,
+ *   BAR, //violation
+ * }
+ * </pre>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;NoEnumTrailingComma&quot;/&gt;
+ * </pre>
+ * <p>
+ * Which results in the following violations:
+ * </p>
+ * <pre>
+ * enum Foo1 {
+ *   FOO,
+ *   BAR; //OK
+ * }
+ * enum Foo2 {
+ *   FOO,
+ *   BAR //OK
+ * }
+ * enum Foo3 {
+ *   FOO,
+ *   BAR, //violation
+ * }
+ * enum Foo4 {
+ *   FOO, // OK
+ *   BAR, // violation
+ *   ;
+ * }
+ * enum Foo5 {
+ *   FOO, // OK
+ *   BAR,; // violation
+ * }
+ * enum Foo6 { FOO, BAR,; } // violation
+ * enum Foo7 { FOO, BAR, } // violation
+ * enum Foo8 {
+ *   FOO, // OK
+ *   BAR // OK
+ *   ;
+ * }
+ * enum Foo9 { FOO, BAR; } // OK
+ * enum Foo10 { FOO, BAR } // OK
+ * </pre>
+ *
+ * @since 8.29
+ */
+@StatelessCheck
+public class NoEnumTrailingCommaCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "no.enum.trailing.comma";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.ENUM_DEF};
```

Comment:
We look at `ENUM_DEF` and examine all children for comma and see if it has another trailing constant def. So we are manually doing our own looping here.

Wouldn't this check be even simplier if we stopped on `ENUM_CONSTANT_DEF` and then only look at the next token or 2? Just check if the next token is a comma it must be followed by another `ENUM_CONSTANT_DEF`?

Reply:
Done